docs/root/configuration/http/http_conn_man/headers.rst

```diff
@@ -9,6 +9,22 @@ is being received) as well as during encoding (when the response is being sent).
 .. contents::
   :local:
 
+.. _config_http_conn_man_headers_scheme:
+
+:scheme
+-------
+
+Envoy will always set the *:scheme* header while processing a request. It should always be available to filters, and should be forwarded upstream for HTTP/1 and HTTP/2, where :ref:`config_http_conn_man_headers_x-forwarded-proto` will be sent for HTTP/1.1
+
+For HTTP/2, and HTTP/3, incoming :scheme headers are trusted and propogated through upstream.
+For HTTP/1, the :scheme header will be set
+1) From the absolute URL if present and valid. An invalid (not "http" or "https") scheme, or an https scheme over an unencrypted connection will result in Envoy rejecting the request.
+2) From the value of the :ref:`config_http_conn_man_headers_x-forwarded-proto` header after sanitization.
```

Comment:
Option 3 is to set from whether the connection is TLS or not? Clarify?

Reply:
XFP is set (if absent) first, scheme copies XFP.
https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L173

Reply:
Here I propose to have another scheme header setting policy:  always set it according to the upstream security level.

Reply:
I'm pretty sure that'll break mesh networks pretty fast, if some hops are secure and some are insecure.
Generally if scheme is set it should be preserved.  If it's not set, it is set according to upstream security level as you suggest.

Reply:
I proposed to add an option in `scheme_header_transformation` to allow setting ":scheme" according to the upstream security level.  I totally agreed with you that scheme should be preserved by the proxy if all proxy in the path works the same way.  But reality is: some are not.  In GCP,  there are GFE,  Cloud-run app-server,  one of them may set scheme wrong.  If we add this option, envoy will have a way to set it correctly.

Reply:
We can add that if necessary but I'd prefer not to unless we actually need to, because that's not what :scheme is for.  As commented, it's a part of the URI and I'm not yet clear on why we'd want to change the URI to one thing for one cluster and another for another.  If you have a use case where a global override doesn't work and we need a per cluster override I can of course look into that.

Reply:
I do also have concerns about consistency if we go the cluster route.  Previously, scheme was only set for outbound traffic so filters didn't see the value.  If we add a "set based on cluster" knob we can best effort set on ingress based on expected cluster, but the expected cluster can change over the course of the filter chain so IMO is much less stable and adds complexity to an already complex process.

That said there was previously support for that, so if we need to again we'll just have to discuss how to make it as clear and consistent as possible.